tensorflow_addons/optimizers/discriminative_layer_training.py

```diff
@@ -0,0 +1,301 @@
+# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Discriminative Layer Training Manager for TensorFlow."""
+
+import tensorflow as tf
+import numpy as np
+from typeguard import typechecked
+import logging
+
+
+class DiscriminativeModelManager:
+    """Class for grouping functions related to model lr_mult management"""
+
+    @staticmethod
+    def _get_layers(layer):
+        """Helper method to access a layer's sublayers as a list or return an empty list
+        """
+        return getattr(layer, "layers", [])
+
+    @staticmethod
+    def _get_lr_mult(layer):
+        """Helper method to access a layer's learning rate multiplier, which defaults to 1 if lr mult is not set
+        """
+        return getattr(layer, "lr_mult", 1.0)
+
+    @staticmethod
+    def _assign_lr_mult(layer, lr_mult):
+        """Helper method to assign a layer's learning rate multiplier, which does nothing if lr mult is already set
+        """
+        if not hasattr(layer, "lr_mult"):
+            layer.lr_mult = lr_mult  # since layer has no lr mult, assign the mult
+            # this method should be called after the user has already assigned some lr mults
+            # to some layers. We just don't want to override any lr mults they assigned
+        else:
+            # we pass here because of propagation to nested layers
+            # users should be able to speficy model.layers[0].layers[0].lr_mult = 0.01
+            # and model.layers[0].lr_mult = 0.1, such that the model.layers[0].layers[0]
+            # keeps its assigned lr mult of 0.01
+            pass
+
+    @staticmethod
+    def _recursively_assign_sublayer_lr_mult(layer):
+
+        """Helper method iterate through all nested layers of an object that behaves like a layer or model
+        By default, we want to propagate the lr mult to the lower layers.
+
+
+        https://stackoverflow.com/questions/6340351/iterating-through-list-of-list-in-python
+        """
+
+        mult = DiscriminativeModelManager._get_lr_mult(layer)
+        layers = DiscriminativeModelManager._get_layers(layer)
+
+        if len(layers) > 0:
```

Comment:
Check for `if layers is not None:`. This is mostly to avoid any weird errors arising out of list mutability. Better safe than sorry.